tests/physical_models/crop/test_leaf_dynamics.py

```diff
@@ -0,0 +1,328 @@
+import copy
+from unittest.mock import patch
+import pytest
+import torch
+import torch.testing
+import yaml
+from numpy.testing import assert_almost_equal
+from pcse.base.parameter_providers import ParameterProvider
+from pcse.engine import Engine
+from pcse.models import Wofost72_PP
+from diffwofost.physical_models.crop.leaf_dynamics import WOFOST_Leaf_Dynamics
+from tests.physical_models.pcse_test_code import EngineTestHelper
+from tests.physical_models.pcse_test_code import WeatherDataProviderTestHelper
+from .. import phy_data_folder
+
+
+def prepare_engine_input(file_path):
+    inputs = yaml.safe_load(open(file_path))
+    agro_management_inputs = inputs["AgroManagement"]
+    cropd = inputs["ModelParameters"]
+
+    weather_data_provider = WeatherDataProviderTestHelper(inputs["WeatherVariables"])
+    crop_model_params_provider = ParameterProvider(cropdata=cropd)
+    external_states = inputs["ExternalStates"]
+
+    # convert parameters to tensors
+    crop_model_params_provider.clear_override()
+    for name in ["SPAN", "TDWI", "TBASE", "PERDL", "RGRLAI"]:
```

Comment:
Why is this list hardcoded? Can't you iterate over the keys in the provider? In that case 2 extra parameters would be iterated over: SLATB and KDIFTB

Reply:
The types of the parameters `SLATB` and `KDIFTB` are `AfgenTrait` and not `float`, see [here](https://github.com/WUR-AI/diffWOFOST/blob/0984aa5f44c40a1e37528ef4f2916233c303364e/src/diffwofost/physical_models/crop/leaf_dynamics.py#L121-L122). This type is used as some sort of lookup table with interpolation, see class definition [here](https://github.com/ajwdewit/pcse/blob/6c22e0345f71338db085c4159dc5fe2463ea1972/pcse/util.py#L604). Since this PR is about two parameters of `"SPAN" ` and `"TDWI"`, only those that effect the gradient path of `"SPAN" ` and `"TDWI"` are converted to tensor in this PR. All the other parameters should be tested in the next iteration.